File: libraries/typescript/src/providers/google-llm.ts
/**
 * Google Gemini LLM provider for Patter's pipeline mode.
 *
 * Implements the ``LLMProvider`` interface against the Gemini Developer
 * API's streaming endpoint (``:streamGenerateContent?alt=sse``).
 * OpenAI-style messages/tools are translated into Gemini's ``contents``
 * and ``tools`` shapes, and streamed response parts are normalised to
 * Patter's ``{ type: 'text' | 'tool_call' | 'done' }`` chunks.
 *
 * Implementation notes:
 *   * Uses native ``fetch`` against the REST SSE endpoint so we don't
 *     pull in a large SDK dependency.
 *   * Single class that satisfies Patter's ``LLMProvider`` interface.
 *   * Vertex AI support (which requires GCP auth) is not included — only
 *     the Developer API (API key) path is supported. Vertex can be added
 *     by a follow-up PR once credential plumbing is in place.
 */

import type { LLMChunk, LLMProvider, LLMStreamOptions } from "../llm-loop";
import { mergeAbortSignals } from "../llm-loop";
import { getLogger } from '../logger';

/** Known Google Gemini chat models. */
export const GoogleModel = {
  GEMINI_2_5_FLASH: 'gemini-2.5-flash',
  GEMINI_2_5_PRO: 'gemini-2.5-pro',
  GEMINI_2_0_FLASH: 'gemini-2.0-flash',
  GEMINI_2_0_FLASH_LITE: 'gemini-2.0-flash-lite',
  GEMINI_1_5_FLASH: 'gemini-1.5-flash',
  GEMINI_1_5_PRO: 'gemini-1.5-pro',
} as const;
/** Union of {@link GoogleModel} string values. */
export type GoogleModel = (typeof GoogleModel)[keyof typeof GoogleModel];

const DEFAULT_MODEL: string = GoogleModel.GEMINI_2_5_FLASH;
const DEFAULT_BASE_URL = 'https://generativelanguage.googleapis.com/v1beta';

/** Constructor options for {@link GoogleLLMProvider}. */
export interface GoogleLLMOptions {
  apiKey: string;
  model?: string;
  baseUrl?: string;
  temperature?: number;
  maxOutputTokens?: number;
}

interface GeminiPart {
  text?: string;
  functionCall?: { name?: string; args?: Record<string, unknown>; id?: string };
  functionResponse?: {
    name?: string;
    response?: Record<string, unknown>;
    id?: string;
  };
}

interface GeminiContent {
  role: 'user' | 'model';
  parts: GeminiPart[];
}

interface OpenAIToolDef {
  type?: string;
  function?: {
    name: string;
    description?: string;
    parameters?: Record<string, unknown>;
  };
  name?: string;
  description?: string;
  parameters?: Record<string, unknown>;
}

/** LLM provider backed by Google Gemini (Developer API, streaming SSE). */
export class GoogleLLMProvider implements LLMProvider {
  private readonly apiKey: string;
  readonly model: string;
  private readonly baseUrl: string;
  private readonly temperature?: number;
  private readonly maxOutputTokens?: number;

  constructor(options: GoogleLLMOptions) {
    if (!options.apiKey) {
      throw new Error(
        'Google API key is required. Pass it via { apiKey } or read GOOGLE_API_KEY from the environment.',
      );
    }
    this.apiKey = options.apiKey;
    this.model = options.model ?? DEFAULT_MODEL;
    this.baseUrl = options.baseUrl ?? DEFAULT_BASE_URL;
    this.temperature = options.temperature;
    this.maxOutputTokens = options.maxOutputTokens;
  }

  /** Stream Patter-format LLM chunks from the Gemini SSE endpoint. */
  async *stream(
    messages: Array<Record<string, unknown>>,
    tools?: Array<Record<string, unknown>> | null,
    opts?: LLMStreamOptions,
  ): AsyncGenerator<LLMChunk, void, unknown> {
    const { systemInstruction, contents } = toGeminiContents(messages);
    const geminiTools = tools ? toGeminiTools(tools as OpenAIToolDef[]) : null;

    const body: Record<string, unknown> = { contents };
    if (systemInstruction) {
      body.systemInstruction = { role: 'system', parts: [{ text: systemInstruction }] };
    }
    if (geminiTools) body.tools = geminiTools;

    const generationConfig: Record<string, unknown> = {};
    if (this.temperature !== undefined) generationConfig.temperature = this.temperature;
    if (this.maxOutputTokens !== undefined)
      generationConfig.maxOutputTokens = this.maxOutputTokens;
    if (Object.keys(generationConfig).length > 0) body.generationConfig = generationConfig;

    const url =
      `${this.baseUrl}/models/${encodeURIComponent(this.model)}:streamGenerateContent?alt=sse` +
      `&key=${encodeURIComponent(this.apiKey)}`;

    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
      signal: mergeAbortSignals(opts?.signal, AbortSignal.timeout(30_000)),
    });

    if (!response.ok) {
      const errText = await response.text();
      getLogger().error(`Gemini API error: ${response.status} ${errText}`);
      return;
    }

    const reader = response.body?.getReader();
    if (!reader) return;

    const decoder = new TextDecoder();
    let buffer = '';
    let nextIndex = 0;
    let lastUsage:
      | { promptTokenCount?: number; candidatesTokenCount?: number; cachedContentTokenCount?: number }
      | undefined;

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split('\n');
      buffer = lines.pop() || '';

      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed.startsWith('data: ')) continue;
        const data = trimmed.slice(6);
        if (!data) continue;

        let payload: {
          candidates?: Array<{
            content?: { parts?: GeminiPart[] };
          }>;
          usageMetadata?: {
            promptTokenCount?: number;
            candidatesTokenCount?: number;
            cachedContentTokenCount?: number;
          };
        };
        try {
          payload = JSON.parse(data);
        } catch {
          continue;
        }

        // Gemini emits usageMetadata on every chunk (running total). Don't
        // yield until the stream is over — otherwise we'd double-count by
        // accumulating partial sums.
        if (payload.usageMetadata) {
          lastUsage = payload.usageMetadata;
        }

        const candidate = payload.candidates?.[0];
        const parts = candidate?.content?.parts ?? [];
        for (const part of parts) {
          if (part.functionCall) {
            const args = part.functionCall.args ?? {};
            const callId =
              part.functionCall.id ?? `gemini_call_${nextIndex}`;
            yield {
              type: 'tool_call',
              index: nextIndex,
              id: callId,
              name: part.functionCall.name ?? '',
              arguments: JSON.stringify(args),
            };
            nextIndex++;
            continue;
          }
          if (part.text) {
            yield { type: 'text', content: part.text };
          }
        }
      }
    }

    if (lastUsage) {
      yield {
        type: 'usage',
        inputTokens: lastUsage.promptTokenCount,
        outputTokens: lastUsage.candidatesTokenCount,
        cacheReadInputTokens: lastUsage.cachedContentTokenCount ?? 0,
      };
    }

    yield { type: 'done' };
  }
}

// ---------------------------------------------------------------------------
// Translation helpers (OpenAI format -> Gemini REST contents)
// ---------------------------------------------------------------------------

function toGeminiTools(tools: OpenAIToolDef[]): Array<Record<string, unknown>> {
  const functionDeclarations = tools.map((t) => {
    const fn = t.function ?? t;
    return {
      name: String(fn.name ?? ''),
      description: String(fn.description ?? ''),
      parameters: fn.parameters ?? { type: 'object', properties: {} },
    };
  });
  if (functionDeclarations.length === 0) return [];
  return [{ functionDeclarations }];
}

interface OpenAIStyleMessage {
  role?: string;
  content?: string | Array<Record<string, unknown>>;
  tool_calls?: Array<{
    id?: string;
    function?: { name?: string; arguments?: string };
  }>;
  tool_call_id?: string;
  name?: string;
}

function toGeminiContents(
  messages: Array<Record<string, unknown>>,
): { systemInstruction: string; contents: GeminiContent[] } {
  const systemParts: string[] = [];
  const contents: GeminiContent[] = [];

  for (const rawMsg of messages as OpenAIStyleMessage[]) {
    const role = rawMsg.role;

    if (role === 'system') {
      if (typeof rawMsg.content === 'string' && rawMsg.content) {
        systemParts.push(rawMsg.content);
      }
      continue;
    }

    if (role === 'user') {
      if (typeof rawMsg.content === 'string' && rawMsg.content) {
        contents.push({ role: 'user', parts: [{ text: rawMsg.content }] });
      }
      continue;
    }

    if (role === 'assistant') {
      const parts: GeminiPart[] = [];
      if (typeof rawMsg.content === 'string' && rawMsg.content) {
        parts.push({ text: rawMsg.content });
      }
      for (const tc of rawMsg.tool_calls ?? []) {
        let args: Record<string, unknown> = {};
        try {
          const parsed = JSON.parse(tc.function?.arguments ?? '{}');
          if (parsed && typeof parsed === 'object') args = parsed as Record<string, unknown>;
        } catch {
          args = {};
        }
        parts.push({
          functionCall: {
            name: tc.function?.name ?? '',
            args,
            id: tc.id,
          },
        });
      }
      if (parts.length > 0) contents.push({ role: 'model', parts });
      continue;
    }

    if (role === 'tool') {
      const raw = rawMsg.content;
      let response: Record<string, unknown>;
      if (typeof raw === 'string') {
        try {
          const parsed = JSON.parse(raw);
          response =
            parsed && typeof parsed === 'object' && !Array.isArray(parsed)
              ? (parsed as Record<string, unknown>)
              : { result: parsed };
        } catch {
          response = { result: raw };
        }
      } else {
        response = (raw as unknown as Record<string, unknown>) ?? {};
      }
      contents.push({
        role: 'user',
        parts: [
          {
            functionResponse: {
              name: rawMsg.name ?? rawMsg.tool_call_id ?? '',
              response,
              id: rawMsg.tool_call_id,
            },
          },
        ],
      });
      continue;
    }
  }

  return { systemInstruction: systemParts.join('\n\n'), contents };
}

File: libraries/python/getpatter/providers/google_llm.py
"""Google Gemini LLM provider for Patter's pipeline mode.

Backed by the ``google-genai`` SDK (``google.genai.Client``) which supports
both the Gemini Developer API and Vertex AI. The provider:

  * Translates OpenAI-formatted messages into Gemini ``Content`` turns
    (including ``function_call`` / ``function_response`` parts) so callers
    don't need a second message format.
  * Maps Gemini stream events (``text`` parts, ``function_call`` parts) to
    the Patter chunk protocol ``{"type": "text"|"tool_call"|"done"}``.
"""

from __future__ import annotations

import asyncio
import json
import logging
import os
from enum import StrEnum
from typing import Any, AsyncIterator

logger = logging.getLogger("getpatter")

__all__ = ["GoogleLLMProvider", "GoogleModel", "GoogleVertexLocation"]


class GoogleModel(StrEnum):
    """Known Google Gemini chat models."""

    GEMINI_2_5_FLASH = "gemini-2.5-flash"
    GEMINI_2_5_PRO = "gemini-2.5-pro"
    GEMINI_2_0_FLASH = "gemini-2.0-flash"
    GEMINI_2_0_FLASH_LITE = "gemini-2.0-flash-lite"
    GEMINI_1_5_FLASH = "gemini-1.5-flash"
    GEMINI_1_5_PRO = "gemini-1.5-pro"


class GoogleVertexLocation(StrEnum):
    """Common Vertex AI region codes accepted via the ``location`` arg."""

    US_CENTRAL1 = "us-central1"
    US_EAST1 = "us-east1"
    US_EAST4 = "us-east4"
    US_WEST1 = "us-west1"
    EUROPE_WEST1 = "europe-west1"
    EUROPE_WEST4 = "europe-west4"
    ASIA_NORTHEAST1 = "asia-northeast1"
    ASIA_SOUTHEAST1 = "asia-southeast1"


_DEFAULT_MODEL = GoogleModel.GEMINI_2_5_FLASH.value
_DEFAULT_VERTEX_LOCATION = GoogleVertexLocation.US_CENTRAL1.value


class GoogleLLMProvider:
    """LLM provider backed by Google Gemini (``google-genai`` SDK).

    Supports both the Gemini Developer API (with an API key) and
    Vertex AI (with Google Cloud credentials).  Streams chunks in
    Patter's ``{"type": "text" | "tool_call" | "done", ...}`` protocol.

    Args:
        api_key: Google API key for the Gemini Developer API. If
            omitted, ``GOOGLE_API_KEY`` is read from the environment.
            Ignored when ``vertexai=True``.
        model: Gemini model ID. Defaults to ``gemini-2.5-flash``.
        vertexai: If True, use Vertex AI instead of the Developer API.
        project: GCP project (Vertex AI only).
        location: GCP location (Vertex AI only). Defaults to
            ``us-central1``.
        temperature: Optional sampling temperature.
        max_output_tokens: Optional output token cap.
    """

    def __init__(
        self,
        api_key: str | None = None,
        model: Union[GoogleModel, str] = _DEFAULT_MODEL,
        vertexai: bool = False,
        project: str | None = None,
        location: str | None = None,
        temperature: float | None = None,
        max_output_tokens: int | None = None,
    ) -> None:
        try:
            from google.genai import Client
        except ImportError as e:
            raise RuntimeError(
                "The 'google-genai' package is required for GoogleLLMProvider. "
                "Install it with: pip install 'getpatter[google]'"
            ) from e

        use_vertexai = (
            vertexai
            if vertexai
            else os.environ.get("GOOGLE_GENAI_USE_VERTEXAI", "0").lower()
            in ["true", "1"]
        )

        resolved_key: str | None = None
        gcp_project = project or os.environ.get("GOOGLE_CLOUD_PROJECT")
        gcp_location = (
            location
            or os.environ.get("GOOGLE_CLOUD_LOCATION")
            or _DEFAULT_VERTEX_LOCATION
        )

        if use_vertexai:
            if not gcp_project:
                raise ValueError(
                    "Project is required for Vertex AI, either via the 'project' "
                    "argument or the GOOGLE_CLOUD_PROJECT environment variable."
                )
        else:
            resolved_key = api_key or os.environ.get("GOOGLE_API_KEY")
            if not resolved_key:
                raise ValueError(
                    "Google API key is required, either as the 'api_key' argument "
                    "or via the GOOGLE_API_KEY environment variable."
                )
            gcp_project = None
            gcp_location = None

        self._client = Client(
            api_key=resolved_key,
            vertexai=use_vertexai,
            project=gcp_project,
            location=gcp_location,
        )
        self._model = model
        self._temperature = temperature
        self._max_output_tokens = max_output_tokens

    async def stream(
        self,
        messages: list[dict],
        tools: list[dict] | None = None,
        *,
        cancel_event: asyncio.Event | None = None,
    ) -> AsyncIterator[dict]:
        """Stream chunks from Gemini's ``generate_content_stream``.

        ``cancel_event`` (set on barge-in by the stream handler) is checked
        between chunks and short-circuits the stream so the underlying
        request is freed immediately instead of blocking the next user
        transcript behind a long-running fetch.
        """
        from google.genai import types

        system_instruction, contents = _to_gemini_contents(messages)
        gemini_tools = _to_gemini_tools(tools) if tools else None

        config_kwargs: dict[str, Any] = {}
        if system_instruction:
            config_kwargs["system_instruction"] = [types.Part(text=system_instruction)]
        if gemini_tools is not None:
            config_kwargs["tools"] = gemini_tools
        if self._temperature is not None:
            config_kwargs["temperature"] = self._temperature
        if self._max_output_tokens is not None:
            config_kwargs["max_output_tokens"] = self._max_output_tokens

        config = types.GenerateContentConfig(**config_kwargs) if config_kwargs else None

        stream = await self._client.aio.models.generate_content_stream(
            model=self._model,
            contents=contents,
            config=config,
        )

        # Gemini does not provide a stable per-function-call index across
        # chunks, so we assign a monotonically increasing index per
        # function_call part that we see.
        next_index = 0

        last_usage = None
        async for response in stream:
            if cancel_event is not None and cancel_event.is_set():
                return
            # Gemini emits ``usage_metadata`` cumulatively on each chunk.
            # Capture only the most recent value so we yield ONE usage
            # event with the final totals (avoids double-counting).
            usage = getattr(response, "usage_metadata", None)
            if usage is not None:
                last_usage = usage

            if not getattr(response, "candidates", None):
                continue
            candidate = response.candidates[0]
            content = getattr(candidate, "content", None)
            if not content or not getattr(content, "parts", None):
                continue

            for part in content.parts:
                function_call = getattr(part, "function_call", None)
                if function_call:
                    args = getattr(function_call, "args", {}) or {}
                    call_id = (
                        getattr(function_call, "id", None)
                        or f"gemini_call_{next_index}"
                    )
                    yield {
                        "type": "tool_call",
                        "index": next_index,
                        "id": call_id,
                        "name": getattr(function_call, "name", "") or "",
                        "arguments": json.dumps(args),
                    }
                    next_index += 1
                    continue

                text = getattr(part, "text", None)
                if text:
                    yield {"type": "text", "content": text}

        if last_usage is not None:
            yield {
                "type": "usage",
                "input_tokens": getattr(last_usage, "prompt_token_count", 0) or 0,
                "output_tokens": getattr(last_usage, "candidates_token_count", 0) or 0,
                "cache_read_tokens": getattr(
                    last_usage, "cached_content_token_count", 0
                )
                or 0,
            }

        yield {"type": "done"}


# ---------------------------------------------------------------------------
# Message / tool translation (OpenAI format -> google.genai types)
# ---------------------------------------------------------------------------


def _to_gemini_tools(tools: list[dict]) -> list[Any]:
    """Convert OpenAI-style tool definitions to Gemini ``Tool`` objects."""
    from google.genai import types

    function_decls: list[types.FunctionDeclaration] = []
    for tool in tools:
        fn = tool.get("function", tool)
        function_decls.append(
            types.FunctionDeclaration(
                name=fn["name"],
                description=fn.get("description", ""),
                parameters=fn.get("parameters", {"type": "object", "properties": {}}),
            )
        )
    if not function_decls:
        return []
    return [types.Tool(function_declarations=function_decls)]


def _to_gemini_contents(messages: list[dict]) -> tuple[str, list[Any]]:
    """Convert OpenAI-style messages to (system_instruction, Gemini Contents).

    Gemini expects:
      * ``system`` as a top-level ``GenerateContentConfig.system_instruction``.
      * ``user`` / ``model`` turns with ``Part`` lists.
      * Tool calls are ``Part(function_call=...)`` on ``model`` turns.
      * Tool results are ``Part(function_response=...)`` on ``user`` turns.
    """
    from google.genai import types

    system_parts: list[str] = []
    contents: list[types.Content] = []

    for msg in messages:
        role = msg.get("role")

        if role == "system":
            content = msg.get("content")
            if isinstance(content, str) and content:
                system_parts.append(content)
            continue

        if role == "user":
            content = msg.get("content", "")
            if isinstance(content, str) and content:
                contents.append(
                    types.Content(role="user", parts=[types.Part(text=content)])
                )
            continue

        if role == "assistant":
            parts: list[types.Part] = []
            text = msg.get("content")
            if isinstance(text, str) and text:
                parts.append(types.Part(text=text))

            for tc in msg.get("tool_calls", []) or []:
                fn = tc.get("function", {})
                try:
                    args = json.loads(fn.get("arguments", "") or "{}")
                except json.JSONDecodeError:
                    args = {}
                parts.append(
                    types.Part(
                        function_call=types.FunctionCall(
                            name=fn.get("name", ""),
                            args=args,
                            id=tc.get("id"),
                        )
                    )
                )

            if parts:
                contents.append(types.Content(role="model", parts=parts))
            continue

        if role == "tool":
            tool_call_id = msg.get("tool_call_id", "")
            raw = msg.get("content", "")
            try:
                response_dict = json.loads(raw) if isinstance(raw, str) else dict(raw)
                if not isinstance(response_dict, dict):
                    response_dict = {"result": response_dict}
            except (json.JSONDecodeError, TypeError):
                response_dict = {"result": raw}
            contents.append(
                types.Content(
                    role="user",
                    parts=[
                        types.Part(
                            function_response=types.FunctionResponse(
                                name=msg.get("name", "") or tool_call_id,
                                response=response_dict,
                                id=tool_call_id or None,
                            )
                        )
                    ],
                )
            )
            continue

    return "\n\n".join(system_parts), contents
